//-------------------------------------------------------------------------------
// Name:        console.proto
// Purpose:     protobuf and grpc for remote control of navigation server
//
// Author:      Laurent Carré
//
// Created:     03/05/2022
// Copyright:   (c) Laurent Carré Sterwen Technology 2021-2025
// Licence:     Eclipse Public License 2.0
//-------------------------------------------------------------------------------

syntax = "proto3";

import "arguments.proto";
import "services_server.proto";



message CouplerMsg {
  string  name=1;
  string coupler_class=2;
  enum Coupler_state {
    STOPPED = 0;
    RUNNING = 1;
    SUSPENDED = 3;
  }
  Coupler_state state=3;
  enum Device_state {
    NOT_READY=0;
    OPEN=1;
    CONNECTED=2;
    ACTIVE=3;
  }
  Device_state dev_state=4;
  string protocol=5;
  uint32 msg_in=6;
  uint32 msg_raw=13;
  uint32 msg_out=7;
  string status=8;
  uint32 error=9;
  float input_rate=10;  // messages / sec
  float input_rate_raw=12;
  float output_rate=11;
}

message ServiceMsg {
  string name = 1;
  string class = 2;
}

message PublisherMsg {
  string name=1;
  string class=2;
  bool active=3;
}

message Request {
  uint32 id=1;
  string cmd=2;
  string target=3;
  ArgumentList kwargs=4;
}

message Response {
  uint32 id=1;
  string status=2;
  ArgumentList response_values=3;
}

service NavigationConsole {
  rpc ServerCmd(Request) returns (Response) {}
  rpc ServerStatus(Request) returns (SystemProcessMsg) {}
  rpc GetCouplers(Request) returns (stream CouplerMsg) {}
  rpc GetCoupler(Request) returns (CouplerMsg) {}
  rpc CouplerCmd(Request) returns (Response) {}
}

